docs/contributor/memory_management.md

```diff
@@ -0,0 +1,177 @@
+
+### Garbage Collection in Compiled Programs
+
+In most Grain programs, the above functions do not need to be invoked by users. Instead, the compiler automatically inserts calls to them where required. When developing in `compcore.re`, it is important to understand the conventions used by the garbage collector.
+
+**Reference counts are incremented upon access.** With some exceptions in generated code, reference counts are incremented whenever a value is accessed (loaded from storage slots such as arguments, locals, and globals). This streamlines the conventions in generated code, as function calls require no specific `incRef` invocation for arguments (see "References are callee-owned" subsection below). For clarity, a couple of notes:
+
+- If a value is stored somewhere, no `incRef`/`decRef` calls are needed (again, there are rare exceptions to this in the compiler, specifically regarding "swap slots", which are local variables used as registers).
+- If the computation no longer needs a live value (e.g. after a statement in the middle of a block), it should be `decRef`ed.
+- No special `incRef` is needed for return values, as the access of those values performs the requisite `incRef`. It should be noted that in generated WebAssembly code, the function postamble looks as follows:
+
+```
+swap_slot <- return_value
+for argument in arguments:
+    decref(argument)
+for local_variable in local_variables:
+    decref(local_variable)
+return load_swap(return_value) # see note below
+```
+
+Note that `load_swap` is one of the aforementioned exceptions which does not `incRef` the value upon access. Swap slots are used by the compiler to temporarily store intermediate values during computation, and they do not need to have their contents `incRef`ed; i.e. since they are only used for intermediate values, any references they hold are treated as weak and do not impact the reference count.
+
+**References are callee-owned.** This has a couple of implications:
+
+- If a function is called with an argument, it is the responsibility of that function to `decRef` that argument
+- If a value is passed to a function with a reference count of `n`, when it returns, it will have the a reference count of `n-1` (unless it was stored in an external location, in which case the reference count may be higher).
+
+**Functions contain references to themselves.** In order to support closures, Grain's compiler emits code which passes the current closure to each function as an extra argument. With respect to the calling convention, this extra argument is the same as normal arguments, so it follows the same `incRef`-before-function-call pattern as standard arguments (and must be `decRef`ed before functions return). This is particularly important to keep in mind when writing `@disableGC` code, as discussed below.
```

Comment:
I'd love to see this info before the `Reference counts are incremented upon access` section! I like having the concepts and then getting into our implementation of it. It might even be better to have two subsections of implementations—`Reference counts are incremented upon access` for how it's implemented in the compiler, and then this other bit below for how it should be implemented in Grain code 😎

Reply:
@ospencer I did the rearrangement, but I am not sure what would be the best way to break this apart into subsections as you describe. Do you want to take another look and see if it now looks like what you'd expect?

Reply:
Looks great!